File: src/music_widget/visualizer/styles.py
"""Visualizer draw functions. Each takes (cr, w, h, bars, accent, teal, state)
and draws onto the cairo context.

`state` is a per-canvas dict the canvas owns; styles that need history (flame)
read/write into it.
"""

import cairo
import math


STYLE_NAMES = ["bars", "wave", "blocks", "flame", "mirror", "dots", "ring"]
STYLE_LABELS = {
    "bars": "Bars",
    "wave": "Wave",
    "blocks": "Blocks",
    "flame": "Flame",
    "mirror": "Mirror",
    "dots": "Dots",
    "ring": "Ring",
}


def draw_bars(cr, w, h, bars, accent, teal, _state):
    n = len(bars) or 1
    bw = w / n
    gap = max(1.0, bw * 0.18)
    for i, v in enumerate(bars):
        bh = max(2.0, v / 100 * h)
        x = i * bw + gap / 2
        g = cairo.LinearGradient(x, h, x, h - bh)
        g.add_color_stop_rgba(0, *accent, 0.9)
        g.add_color_stop_rgba(1, *teal, 0.9)
        cr.set_source(g)
        cr.rectangle(x, h - bh, bw - gap, bh)
        cr.fill()


def draw_wave(cr, w, h, bars, accent, teal, _state):
    n = len(bars)
    if n < 2:
        return
    mid = h / 2
    pts = [(i / (n - 1) * w, mid - bars[i] / 100 * mid * 0.85) for i in range(n)]

    def curve(pts_in, col, a):
        cr.set_source_rgba(*col, a)
        cr.set_line_width(2.0)
        cr.move_to(*pts_in[0])
        for j in range(1, n - 2):
            cx = (pts_in[j][0] + pts_in[j + 1][0]) / 2
            cy = (pts_in[j][1] + pts_in[j + 1][1]) / 2
            cr.curve_to(pts_in[j][0], pts_in[j][1], cx, cy, cx, cy)
        cr.line_to(*pts_in[-1])
        cr.stroke()

    curve(pts, accent, 0.9)
    curve([(x, mid + (mid - y)) for x, y in pts], teal, 0.45)


def draw_blocks(cr, w, h, bars, accent, teal, _state):
    n = len(bars) or 1
    bw = w / n
    bh, step = 5, 7
    for i, v in enumerate(bars):
        th = max(bh, v / 100 * h)
        x = i * bw + 1
        y = h - step
        while y > h - th:
            t = (h - y) / h
            r = accent[0] * (1 - t) + teal[0] * t
            g = accent[1] * (1 - t) + teal[1] * t
            b = accent[2] * (1 - t) + teal[2] * t
            cr.set_source_rgba(r, g, b, 0.85)
            cr.rectangle(x, y, bw - 2, bh)
            cr.fill()
            y -= step


_FLAME_ROWS = 20


def draw_flame(cr, w, h, bars, _accent, _teal, state):
    n = len(bars) or 1
    grid = state.get("flame")
    if grid is None or len(grid[0]) != n:
        grid = [[0.0] * n for _ in range(_FLAME_ROWS)]
        state["flame"] = grid

    rows = _FLAME_ROWS
    for r in range(rows - 1):
        for c in range(n):
            left = grid[r + 1][max(0, c - 1)]
            mid = grid[r + 1][c]
            right = grid[r + 1][min(n - 1, c + 1)]
            grid[r][c] = (left + mid * 2 + right) / 4 * 0.93
    for c in range(n):
        grid[rows - 1][c] = bars[c] / 100

    bw, rh = w / n, h / rows
    for row in range(rows):
        for col in range(n):
            v = grid[row][col]
            if v < 0.04:
                continue
            if v < 0.35:
                t = v / 0.35
                clr = (0.7 * t, 0.05 * t, 0, v)
            elif v < 0.65:
                t = (v - 0.35) / 0.30
                clr = (0.7, 0.05 + 0.45 * t, 0, v)
            else:
                t = (v - 0.65) / 0.35
                clr = (0.9, 0.5 + 0.3 * t, 0.1 * t, min(1, v))
            cr.set_source_rgba(*clr)
            cr.rectangle(col * bw, (rows - 1 - row) * rh, bw + 0.5, rh + 0.5)
            cr.fill()


def draw_mirror(cr, w, h, bars, accent, teal, _state):
    """Bars mirrored from the horizontal midline — classic spectrum analyzer."""
    n = len(bars) or 1
    bw = w / n
    gap = max(1.0, bw * 0.18)
    mid = h / 2
    for i, v in enumerate(bars):
        half = max(1.5, v / 100 * mid * 0.95)
        x = i * bw + gap / 2
        # Upper half — accent on top, teal at midline
        g_up = cairo.LinearGradient(x, mid, x, mid - half)
        g_up.add_color_stop_rgba(0, *teal, 0.85)
        g_up.add_color_stop_rgba(1, *accent, 0.9)
        cr.set_source(g_up)
        cr.rectangle(x, mid - half, bw - gap, half)
        cr.fill()
        # Lower half — symmetric
        g_dn = cairo.LinearGradient(x, mid, x, mid + half)
        g_dn.add_color_stop_rgba(0, *teal, 0.85)
        g_dn.add_color_stop_rgba(1, *accent, 0.9)
        cr.set_source(g_dn)
        cr.rectangle(x, mid, bw - gap, half)
        cr.fill()


def draw_dots(cr, w, h, bars, accent, teal, _state):
    """A single filled dot per band at its current peak height."""
    n = len(bars) or 1
    bw = w / n
    r_max = max(2.5, bw * 0.35)
    for i, v in enumerate(bars):
        t = max(0.0, min(1.0, v / 100))
        # Color lerps from teal (quiet) to accent (loud)
        r = teal[0] * (1 - t) + accent[0] * t
        g = teal[1] * (1 - t) + accent[1] * t
        b = teal[2] * (1 - t) + accent[2] * t
        radius = max(2.0, r_max * (0.4 + 0.6 * t))
        cx = i * bw + bw / 2
        cy = h - max(radius + 2, t * (h - radius - 2))
        cr.set_source_rgba(r, g, b, 0.55 + 0.45 * t)
        cr.arc(cx, cy, radius, 0, 2 * math.pi)
        cr.fill()


def draw_ring(cr, w, h, bars, accent, teal, _state):
    """Bars radiating from a central ring."""
    n = len(bars) or 1
    cx, cy = w / 2, h / 2
    r_inner = min(w, h) * 0.18
    r_max = min(w, h) / 2 - 4
    for i, v in enumerate(bars):
        angle = (i / n) * 2 * math.pi - math.pi / 2  # start from top
        t = max(0.0, min(1.0, v / 100))
        length = max(2.0, t * (r_max - r_inner))
        x0 = cx + math.cos(angle) * r_inner
        y0 = cy + math.sin(angle) * r_inner
        x1 = cx + math.cos(angle) * (r_inner + length)
        y1 = cy + math.sin(angle) * (r_inner + length)
        # Per-spoke gradient teal → accent outward
        g = cairo.LinearGradient(x0, y0, x1, y1)
        g.add_color_stop_rgba(0, *teal, 0.85)
        g.add_color_stop_rgba(1, *accent, 0.95)
        cr.set_source(g)
        cr.set_line_width(max(2.0, 2 * math.pi * r_inner / n * 0.6))
        cr.move_to(x0, y0)
        cr.line_to(x1, y1)
        cr.stroke()
    # Subtle inner ring
    cr.set_source_rgba(*accent, 0.4)
    cr.set_line_width(1.5)
    cr.arc(cx, cy, r_inner, 0, 2 * math.pi)
    cr.stroke()


DRAWERS = {
    "bars": draw_bars,
    "wave": draw_wave,
    "blocks": draw_blocks,
    "flame": draw_flame,
    "mirror": draw_mirror,
    "dots": draw_dots,
    "ring": draw_ring,
}

File: modules/visualizer/styles.js
// Visualizer draw styles, ported from the cairo originals (styles.py).
// Each drawer takes (ctx, w, h, bars, accent, teal, state) where bars are
// floats in [0,1] and accent/teal are QML color values. `state` is a
// per-canvas object for styles needing history (flame, peak-hold).
.pragma library

var STYLE_NAMES = ["bars", "wave", "blocks", "flame", "mirror", "dots", "ring"];

function rgba(c, a) {
    return "rgba(" + Math.round(c.r * 255) + "," + Math.round(c.g * 255) + ","
        + Math.round(c.b * 255) + "," + a + ")";
}

function lerpRgba(c1, c2, t, a) {
    return "rgba("
        + Math.round((c1.r * (1 - t) + c2.r * t) * 255) + ","
        + Math.round((c1.g * (1 - t) + c2.g * t) * 255) + ","
        + Math.round((c1.b * (1 - t) + c2.b * t) * 255) + "," + a + ")";
}

// Decaying per-band peak markers shared by bars/mirror.
function updatePeaks(state, key, bars) {
    var peaks = state[key];
    if (!peaks || peaks.length !== bars.length)
        peaks = new Array(bars.length).fill(0);
    for (var i = 0; i < bars.length; i++)
        peaks[i] = Math.max(peaks[i] - 0.008, bars[i]);
    state[key] = peaks;
    return peaks;
}

function drawBars(ctx, w, h, bars, accent, teal, state, peakHold) {
    var n = bars.length || 1;
    var bw = w / n;
    var gap = Math.max(1, bw * 0.18);
    var peaks = peakHold ? updatePeaks(state, "peaks", bars) : null;
    for (var i = 0; i < n; i++) {
        var bh = Math.max(2, bars[i] * h);
        var x = i * bw + gap / 2;
        var g = ctx.createLinearGradient(x, h, x, h - bh);
        g.addColorStop(0, rgba(accent, 0.9));
        g.addColorStop(1, rgba(teal, 0.9));
        ctx.fillStyle = g;
        ctx.fillRect(x, h - bh, bw - gap, bh);
        if (peaks && peaks[i] > 0.02) {
            ctx.fillStyle = rgba(teal, 0.9);
            ctx.fillRect(x, h - peaks[i] * h - 2, bw - gap, 2);
        }
    }
}

function drawWave(ctx, w, h, bars, accent, teal) {
    var n = bars.length;
    if (n < 4)
        return;
    var mid = h / 2;
    var pts = [];
    for (var i = 0; i < n; i++)
        pts.push([i / (n - 1) * w, mid - bars[i] * mid * 0.85]);

    function curve(p, col, a) {
        ctx.strokeStyle = rgba(col, a);
        ctx.lineWidth = 2;
        ctx.beginPath();
        ctx.moveTo(p[0][0], p[0][1]);
        for (var j = 1; j < n - 2; j++) {
            var cx = (p[j][0] + p[j + 1][0]) / 2;
            var cy = (p[j][1] + p[j + 1][1]) / 2;
            ctx.quadraticCurveTo(p[j][0], p[j][1], cx, cy);
        }
        ctx.lineTo(p[n - 1][0], p[n - 1][1]);
        ctx.stroke();
    }

    curve(pts, accent, 0.9);
    var mirrored = pts.map(function (pt) {
        return [pt[0], mid + (mid - pt[1])];
    });
    curve(mirrored, teal, 0.45);
}

function drawBlocks(ctx, w, h, bars, accent, teal) {
    var n = bars.length || 1;
    var bw = w / n;
    var bh = 5, step = 7;
    for (var i = 0; i < n; i++) {
        var th = Math.max(bh, bars[i] * h);
        var x = i * bw + 1;
        var y = h - step;
        while (y > h - th) {
            var t = (h - y) / h;
            ctx.fillStyle = lerpRgba(accent, teal, t, 0.85);
            ctx.fillRect(x, y, bw - 2, bh);
            y -= step;
        }
    }
}

var FLAME_ROWS = 20;

function drawFlame(ctx, w, h, bars, accent, teal, state) {
    var n = bars.length || 1;
    var grid = state.flame;
    if (!grid || grid[0].length !== n) {
        grid = [];
        for (var r0 = 0; r0 < FLAME_ROWS; r0++)
            grid.push(new Array(n).fill(0));
        state.flame = grid;
    }

    for (var r = 0; r < FLAME_ROWS - 1; r++) {
        for (var c = 0; c < n; c++) {
            var left = grid[r + 1][Math.max(0, c - 1)];
            var mid = grid[r + 1][c];
            var right = grid[r + 1][Math.min(n - 1, c + 1)];
            grid[r][c] = (left + mid * 2 + right) / 4 * 0.93;
        }
    }
    for (var c2 = 0; c2 < n; c2++)
        grid[FLAME_ROWS - 1][c2] = bars[c2];

    var bw = w / n, rh = h / FLAME_ROWS;
    for (var row = 0; row < FLAME_ROWS; row++) {
        for (var col = 0; col < n; col++) {
            var v = grid[row][col];
            if (v < 0.04)
                continue;
            var clr;
            if (v < 0.35) {
                var t1 = v / 0.35;
                clr = "rgba(" + Math.round(0.7 * t1 * 255) + "," + Math.round(0.05 * t1 * 255) + ",0," + v + ")";
            } else if (v < 0.65) {
                var t2 = (v - 0.35) / 0.30;
                clr = "rgba(179," + Math.round((0.05 + 0.45 * t2) * 255) + ",0," + v + ")";
            } else {
                var t3 = (v - 0.65) / 0.35;
                clr = "rgba(230," + Math.round((0.5 + 0.3 * t3) * 255) + "," + Math.round(0.1 * t3 * 255) + "," + Math.min(1, v) + ")";
            }
            ctx.fillStyle = clr;
            ctx.fillRect(col * bw, (FLAME_ROWS - 1 - row) * rh, bw + 0.5, rh + 0.5);
        }
    }
}

function drawMirror(ctx, w, h, bars, accent, teal, state, peakHold) {
    var n = bars.length || 1;
    var bw = w / n;
    var gap = Math.max(1, bw * 0.18);
    var mid = h / 2;
    var peaks = peakHold ? updatePeaks(state, "peaksM", bars) : null;
    for (var i = 0; i < n; i++) {
        var half = Math.max(1.5, bars[i] * mid * 0.95);
        var x = i * bw + gap / 2;
        var gUp = ctx.createLinearGradient(x, mid, x, mid - half);
        gUp.addColorStop(0, rgba(teal, 0.85));
        gUp.addColorStop(1, rgba(accent, 0.9));
        ctx.fillStyle = gUp;
        ctx.fillRect(x, mid - half, bw - gap, half);
        var gDn = ctx.createLinearGradient(x, mid, x, mid + half);
        gDn.addColorStop(0, rgba(teal, 0.85));
        gDn.addColorStop(1, rgba(accent, 0.9));
        ctx.fillStyle = gDn;
        ctx.fillRect(x, mid, bw - gap, half);
        if (peaks && peaks[i] > 0.02) {
            var ph = peaks[i] * mid * 0.95;
            ctx.fillStyle = rgba(accent, 0.9);
            ctx.fillRect(x, mid - ph - 2, bw - gap, 1.5);
            ctx.fillRect(x, mid + ph + 0.5, bw - gap, 1.5);
        }
    }
}

function drawDots(ctx, w, h, bars, accent, teal) {
    var n = bars.length || 1;
    var bw = w / n;
    var rMax = Math.max(2.5, bw * 0.35);
    for (var i = 0; i < n; i++) {
        var t = Math.max(0, Math.min(1, bars[i]));
        var radius = Math.max(2, rMax * (0.4 + 0.6 * t));
        var cx = i * bw + bw / 2;
        var cy = h - Math.max(radius + 2, t * (h - radius - 2));
        ctx.fillStyle = lerpRgba(teal, accent, t, 0.55 + 0.45 * t);
        ctx.beginPath();
        ctx.arc(cx, cy, radius, 0, 2 * Math.PI);
        ctx.fill();
    }
}

function drawRing(ctx, w, h, bars, accent, teal) {
    var n = bars.length || 1;
    var cx = w / 2, cy = h / 2;
    var rInner = Math.min(w, h) * 0.18;
    var rMax = Math.min(w, h) / 2 - 4;
    ctx.lineWidth = Math.max(2, 2 * Math.PI * rInner / n * 0.6);
    for (var i = 0; i < n; i++) {
        var angle = (i / n) * 2 * Math.PI - Math.PI / 2;
        var t = Math.max(0, Math.min(1, bars[i]));
        var length = Math.max(2, t * (rMax - rInner));
        var x0 = cx + Math.cos(angle) * rInner;
        var y0 = cy + Math.sin(angle) * rInner;
        var x1 = cx + Math.cos(angle) * (rInner + length);
        var y1 = cy + Math.sin(angle) * (rInner + length);
        var g = ctx.createLinearGradient(x0, y0, x1, y1);
        g.addColorStop(0, rgba(teal, 0.85));
        g.addColorStop(1, rgba(accent, 0.95));
        ctx.strokeStyle = g;
        ctx.beginPath();
        ctx.moveTo(x0, y0);
        ctx.lineTo(x1, y1);
        ctx.stroke();
    }
    ctx.strokeStyle = rgba(accent, 0.4);
    ctx.lineWidth = 1.5;
    ctx.beginPath();
    ctx.arc(cx, cy, rInner, 0, 2 * Math.PI);
    ctx.stroke();
}

function draw(style, ctx, w, h, bars, accent, teal, state, peakHold) {
    if (!bars || bars.length === 0)
        return;
    if (style === "wave")
        drawWave(ctx, w, h, bars, accent, teal);
    else if (style === "blocks")
        drawBlocks(ctx, w, h, bars, accent, teal);
    else if (style === "flame")
        drawFlame(ctx, w, h, bars, accent, teal, state);
    else if (style === "mirror")
        drawMirror(ctx, w, h, bars, accent, teal, state, peakHold);
    else if (style === "dots")
        drawDots(ctx, w, h, bars, accent, teal);
    else if (style === "ring")
        drawRing(ctx, w, h, bars, accent, teal);
    else
        drawBars(ctx, w, h, bars, accent, teal, state, peakHold);
}
